### PR TITLE
Fix #1754 Re-enable intro activity tests

### LINKED ISSUE #1754: Re-enable IntroActivity tests
**Type of Test**

* [ ] Unit Test
* [x] UI Test

**Current Status of the tests**
On CI this test would sometimes fail.
This was generally not being able to swipe the view and see the views on the other page.

Operating theory is that a system dialog was interfering at times, preventing view interaction

### PR DESCRIPTION
**Pull request by yashk2000**
Fixes #1754 

Re-enable the intro activity tests.

### CODE REVIEW

**Comment by yashk2000** [comment]
@macgills I ran 5 tests these are the results. How should I proceed now? The errors just seem to be random and are not happening when I run them on my laptop. 

| Test number | Pass/Fail | API Level for which failure happened | Log                                                                                                                                                                                                           |
|-------------|-----------|--------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 1           | Pass      | -                                    | -                                                                                                                                                                                                             |
| 2           | Fail      | 21                                   | There was 1 failure:  1) viewIsSwipeableAndNavigatesToMain(org.kiwix.kiwixmobile.intro.IntroActivityTest)  java.lang.RuntimeException: No view found with Id org.kiwix.kiwixmobile:id/get_started             |
| 3           | Fail      | 29                                   | There was 1 failure:  1) viewIsSwipeableAndNavigatesToMain(org.kiwix.kiwixmobile.intro.IntroActivityTest)  java.lang.RuntimeException: java.lang.RuntimeException: No view found with text Save books offline |
| 4           | Pass      | -                                    | -                                                                                                                                                                                                             |
| 5           | Fail      | 21                                   | There was 1 failure:   1) viewIsSwipeableAndNavigatesToMain(org.kiwix.kiwixmobile.intro.IntroActivityTest)   java.lang.RuntimeException: No view found with Id org.kiwix.kiwixmobile:id/get_started           |

**Comment by macgills** [comment]
This was the reason the ticket was made. It is failing for totally unknown reasons, perhaps these emulators have an issue specific to them where somehow an unknown dialog shows itself. You could look into the complicated process of taking a screenshot of the device and then uploading it with [upload_artifact](https://github.com/actions/upload-artifact) but other than that I am not sure how to get more information from this.

I have tried increasing timeouts, a variety of different methods for swiping the viewpager and a lot of other things. Testing on emulators is just a bit crap

**Comment by yashk2000** [comment]
@macgills how will the screenshots help in providing more information?
When I run the emulator tests locally, they pass and I don't see any unknown dialog showing up on the emulator.

**Comment by macgills** [comment]
> @macgills how will the screenshots help in providing more information?
> When I run the emulator tests locally, they pass and I don't see any unknown dialog showing up on the emulator.

Because "something" is happening when these tests get run on the CI and as much as I wish I had access to some sort of powerful scrying orb I don't so the next best thing is automated screenshots on a test failure.

I attempted a lame version with `applyWithViewHierarchyPrinting` but that didn't work incredibly well as it just confirmed that "Yup the views are there, up yours I'm failing anyways"

**Comment by yashk2000** [comment]
@macgills this issue seems to be very irksome judging by your past experiences with it :sweat_smile: 
I'll look into how to use upload_artifact and update the pr with it. Let's hope that will give some positive insights.

**Comment by macgills** [comment]
It makes me want to take a sledgehammer to my screen and oh I don't know live in the woods eating moss or something.

Best of luck

**Comment by yashk2000** [comment]
@macgills could you please view the changes I made. I just intended to take a screenshot if the build fails. Could you please verify if that is what will happen?

I'm not able to find where it is going wrong. I force pushed and triggered the tests repeated 4 times with the current changes in the configuration, but now the tests are not failing. I checked the logs, and the UI tests seem to be completing properly, but I'm not sure.

**Comment by macgills** [comment]
@yashk2000 keep pushing, they'll eventually fail, this is the pain of flaky tests

**Comment by yashk2000** [comment]
@macgills uploading artifcats to GitHub actions works now. But I'm not able to extract screenshots for API levels 21 and 22.

File: contrib/instrumentation.sh [hunk 1]
```diff
@@ -3,3 +3,7 @@
 adb logcat -c
 adb logcat *:E -v color &
 ./gradlew jacocoInstrumentationTestReport
+adb remount
```
Comment (macgills) [inline contrib/instrumentation.sh:6]: Screenshots are overwriting each other for this workflow run as they all have the same name so we can't get a specific image for a failure. If we can pass the api level to the bash script/uploader so the filenames are eg `screencap-21` then we can assess if we are getting useful screenshots. If these screenshots are not useful then me may have to move the screenshot taking internal to the code much like `applyViewHeirarchyPrinting`

Also guessing from the error messages [this](https://stackoverflow.com/questions/40488317/adb-error-remote-object-data-data-com-me-myproject-does-not-exist) might work for API 21/22?

Comment (yashk2000) [reply contrib/instrumentation.sh]: @macgills I wasn't able to get the screenshots in API 21 and 22. I tried what that link had, but I'm already pulling the screenshots from the `sdcard` directory.

Comment (macgills) [reply contrib/instrumentation.sh]: I don't think I saw a commit with this line
`run-as com.yourPackageName`

and when you were trying to use `/mnt/sdcard/` you were copying to `/sdcard/` first which is the problematic directory.

This is what I remember seeing so I could be inaccurate

File: contrib/instrumentation.sh [hunk 2]
```diff
@@ -3,3 +3,6 @@
 adb logcat -c
 adb logcat *:E -v color &
 ./gradlew jacocoInstrumentationTestReport
+adb shell screencap -p /sdcard/screencap.png
```
Comment (macgills) [inline contrib/instrumentation.sh:6]: it looks like it is possible to skip writing it to the sdcard, [link](https://stackoverflow.com/a/37191719/12225251)

File: .github/workflows/coverage.yml [hunk 3]
```diff
@@ -31,6 +31,12 @@ jobs:
           arch: x86_64
           script: bash contrib/instrumentation.sh
 
+      - name: Upload screenshot result
+        uses: actions/upload-artifact@v1
```
Comment (macgills) [inline .github/workflows/coverage.yml:35]: v2 of this was released and it sounds like it sucks waaaaaaaay less

File: contrib/instrumentation.sh (hunk 2 above)
Comment (yashk2000) [reply contrib/instrumentation.sh:6]: Thanks, I'll try this out.

File: .github/workflows/coverage.yml (hunk 3 above)
Comment (yashk2000) [reply .github/workflows/coverage.yml:35]: I'll update it.

**Comment by yashk2000** [comment]
@macgills done. But sadly no test failed. I'll force push a few times.

File: .github/workflows/coverage.yml [hunk 4]
```diff
@@ -31,6 +31,12 @@ jobs:
           arch: x86_64
           script: bash contrib/instrumentation.sh
 
+      - name: Upload screenshot result
```
Comment (macgills) [inline .github/workflows/coverage.yml:34]: we need an `if failure` for this to run when the test above fails

Comment (macgills) [reply .github/workflows/coverage.yml:34]: or `if always` to just do it whenever

**Comment by yashk2000** [comment]
@macgills tests seem to be getting canceled for API 23 repeatedly now. The instrumentation tests succeed but, then the further steps get canceled.

**Comment by yashk2000** [comment]
@macgills how should we proceed with this now?

**Comment by macgills** [comment]
image


@yashk2000 there was 1 pass so I really don't know, keep building and see if this is persistent?

**Comment by yashk2000** [comment]
@macgills sure, I'll trigger the build again.

**Comment by yashk2000** [comment]
@macgills test canceled for PI 24 now :(

**Comment by yashk2000** [comment]
@macgills stuck on test for API 24 for a few hours now. I'm not able to figure out why this is happening.

**Comment by macgills** [comment]
Approach it methodically, identify which change caused these cancellation errors. You could try re-ignoring the tests, reducing the api levels the tests run on.

I believe I had problems from not specifying an _exact_ version for a GitHub action before so try specifying an exact version, downgrading the version, checking the issues on their GitHub repo, opening issues if necessary.

The android emulator action has had numerous upgrades since I last checked so maybe that will help? Now supports down to api 15(!)

**Comment by macgills** [comment]
Also don't worry too much, this ticket is pretty ambitious because CI is frustrating and terrible. I can't say how many weeks of my life Travis, Jenkins & actions have taken from me over the last few years

**Comment by yashk2000** [comment]
@macgills downgrading the artifacts version seemed to do something. I'll force push a couple more times to see if it working properly now.

**Comment by kelson42** [comment]
@yashk2000 @macgills What should we do now? What is the status?

**Comment by macgills** [comment]
@kelson42 I think the status is we still don't know what is happening.

**Comment by yashk2000** [comment]
@macgills what should we do now then?

**Comment by macgills** [comment]
@yashk2000 have you tried upgrading `reactivecircus/android-emulator-runner@v2.0.0`?

If we can secure 3 consecutive green builds then I am fine to reenable IntroActivityTest, will just have to monitor it

**Comment by yashk2000** [comment]
@macgills tests passed 3 times consecutively!!

**Comment by macgills** [comment]
Then lets merge and cross our fingers. Good work on this, deeply frustrating task

**Pull request merged**